// @pandorajs/component-actuator-server/build/src/ComponentActuatorServer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
const component_decorator_1 = require("@pandorajs/component-decorator");
const EndPointManager_1 = require("./EndPointManager");
const ActuatorRestServer_1 = require("./ActuatorRestServer");
const dollar_1 = require("@pandorajs/dollar");
let ComponentActuatorServer = class ComponentActuatorServer {
    constructor(ctx) {
        this.ctx = ctx;
        this.actuatorRestServer = new ActuatorRestServer_1.ActuatorRestServer(ctx);
        ctx.endPointManager = new EndPointManager_1.EndPointManager(this.actuatorRestServer);
    }
    async startAtSupervisor() {
        await this.actuatorRestServer.start();
        if (this.actuatorRestServer.server) {
            const { address, port } = this.actuatorRestServer.server.address();
            dollar_1.consoleLogger.info(`Actuator restful server started at http://${address}:${port}`);
        }
    }
    async stopAtSupervisor() {
        await this.actuatorRestServer.stop();
    }
};
ComponentActuatorServer = __decorate([
    component_decorator_1.componentName('actuatorServer'),
    component_decorator_1.componentConfig({
        actuatorServer: {
            http: {
                enabled: true,
                host: '127.0.0.1',
                port: 7002,
            },
        },
    })
], ComponentActuatorServer);
exports.default = ComponentActuatorServer;
__exportStar(require("./types"), exports);
__exportStar(require("./ActuatorRestServer"), exports);
__exportStar(require("./EndPointManager"), exports);
