// @pandorajs/hub/build/src/const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PANDORA_HUB_CONFIG_MANAGER = exports.OBJECT_ACTION_UNSUBSCRIBE = exports.OBJECT_ACTION_SUBSCRIBE = exports.OBJECT_ACTION_INTROSPECT = exports.OBJECT_ACTION_GET_PROPERTY = exports.OBJECT_ACTION_INVOKE = exports.TIMEOUT_OF_RESPONSE = exports.PANDORA_HUB_ACTION_UNPUBLISH_UP = exports.PANDORA_HUB_ACTION_PUBLISH_UP = exports.PANDORA_HUB_ACTION_OFFLINE_UP = exports.PANDORA_HUB_ACTION_ONLINE_UP = exports.PANDORA_HUB_ACTION_MSG_DOWN = exports.PANDORA_HUB_ACTION_MSG_UP = exports.HUB_SOCKET_NAME = void 0;
exports.HUB_SOCKET_NAME = 'pandora_msg_hub';
exports.PANDORA_HUB_ACTION_MSG_UP = 'PANDORA_HUB_ACTION_MSG_UP';
exports.PANDORA_HUB_ACTION_MSG_DOWN = 'PANDORA_HUB_ACTION_MSG_DOWN';
exports.PANDORA_HUB_ACTION_ONLINE_UP = 'PANDORA_HUB_ACTION_ONLINE_UP';
// export const PANDORA_HUB_ACTION_ONLINE_DOWN = 'PANDORA_HUB_ACTION_ONLINE_DOWN';
exports.PANDORA_HUB_ACTION_OFFLINE_UP = 'PANDORA_HUB_ACTION_OFFLINE_UP';
// export const PANDORA_HUB_ACTION_OFFLINE_DOWN = 'PANDORA_HUB_ACTION_OFFLINE_DOWN';
exports.PANDORA_HUB_ACTION_PUBLISH_UP = 'PANDORA_HUB_ACTION_PUBLISH_UP';
// export const PANDORA_HUB_ACTION_PUBLISH_DOWN = 'PANDORA_HUB_ACTION_PUBLISH_DOWN';
exports.PANDORA_HUB_ACTION_UNPUBLISH_UP = 'PANDORA_HUB_ACTION_UNPUBLISH_UP';
// export const PANDORA_HUB_ACTION_UNPUBLISH_DOWN = 'PANDORA_HUB_ACTION_UNPUBLISH_DOWN';
// export const PANDORA_HUB_ACTION_DISCOVER_UP = 'PANDORA_HUB_ACTION_DISCOVER_UP';
// export const PANDORA_HUB_ACTION_DISCOVER_DOWM = 'PANDORA_HUB_ACTION_DISCOVER_DOWM';
// export const PANDORA_HUB_ACTION_LOOKUP_UP = 'PANDORA_HUB_ACTION_LOOKUP_UP';
// export const PANDORA_HUB_ACTION_LOOKUP_DOWM = 'PANDORA_HUB_ACTION_LOOKUP_DOWM';
exports.TIMEOUT_OF_RESPONSE = 5000;
exports.OBJECT_ACTION_INVOKE = 'OBJECT_ACTION_INVOKE';
exports.OBJECT_ACTION_GET_PROPERTY = 'OBJECT_ACTION_GET_PROPERTY';
exports.OBJECT_ACTION_INTROSPECT = 'OBJECT_ACTION_INTROSPECT';
exports.OBJECT_ACTION_SUBSCRIBE = 'OBJECT_ACTION_SUBSCRIBE';
exports.OBJECT_ACTION_UNSUBSCRIBE = 'OBJECT_ACTION_UNSUBSCRIBE';
exports.PANDORA_HUB_CONFIG_MANAGER = '__PANDORA_HUB_CONFIG_MANAGER';
