// @pandorajs/component-decorator/build/src/ComponentReflector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.componentConfig = exports.componentName = exports.dependencies = exports.ComponentReflector = void 0;
require("reflect-metadata");
class ComponentReflector {
    static getDependencies(klass) {
        return Reflect.getMetadata(ComponentReflector.COMPONENT_DEPENDENCIES, klass);
    }
    static getComponentName(klass) {
        return Reflect.getMetadata(ComponentReflector.COMPONENT_NAME, klass);
    }
    static getComponentConfig(klass) {
        return Reflect.getMetadata(ComponentReflector.COMPONENT_CONFIG, klass);
    }
}
exports.ComponentReflector = ComponentReflector;
ComponentReflector.COMPONENT_NAME = 'COMPONENT_NAME';
ComponentReflector.COMPONENT_DEPENDENCIES = 'COMPONENT_DEPENDENCIES';
ComponentReflector.COMPONENT_CONFIG = 'COMPONENT_CONFIG';
function dependencies(deps) {
    return function (target) {
        Reflect.defineMetadata(ComponentReflector.COMPONENT_DEPENDENCIES, deps, target);
    };
}
exports.dependencies = dependencies;
function componentName(name) {
    return function (target) {
        Reflect.defineMetadata(ComponentReflector.COMPONENT_NAME, name, target);
    };
}
exports.componentName = componentName;
function componentConfig(config) {
    return function (target) {
        Reflect.defineMetadata(ComponentReflector.COMPONENT_CONFIG, config, target);
    };
}
exports.componentConfig = componentConfig;
