// @pandorajs/messenger/build/src/index.js
// eslint-disable-next-line @typescript-eslint/triple-slash-reference
/// <reference path="./sdk-base.d.ts" />
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const client_1 = require("./client");
const server_1 = require("./server");
const assert = require("assert");
const eventName_1 = require("./eventName");
const Clients = new Map();
const Servers = new Map();
function factory(options, type) {
    assert(options.name, 'options.name is required');
    const isClient = type === 'client';
    const MessengerClass = isClient ? client_1.default : server_1.default;
    const key = options.name;
    const pool = isClient ? Clients : Servers;
    if (!pool[key]) {
        pool[key] = new MessengerClass(options);
    }
    return pool[key];
}
var client_2 = require("./client");
Object.defineProperty(exports, "MessengerClient", { enumerable: true, get: function () { return client_2.default; } });
var server_2 = require("./server");
Object.defineProperty(exports, "MessengerServer", { enumerable: true, get: function () { return server_2.default; } });
exports.default = {
    eventName: eventName_1.default,
    Client: client_1.default,
    Server: server_1.default,
    getClient(options) {
        options = Object.assign({ reConnectTimes: 10 }, options);
        return factory(options, 'client');
    },
    getServer(options) {
        return factory(options, 'server');
    },
};
